WizBox API

minhmo1620 opened the pull request:
WizBox is a feature of WizFood website, which aims to create a place for everyone to add their food recommendations and vote for what they like to eat in a group hanging out. Think about WizBox as a room for everyone to gather, everyone can add their recommendation through ```/options``` and then vote for what they want to eat. When voting, each user will have 3 choices for each option and the choices are indexed through request ( 0 - sad, 1 - neutral, 2 - happy)

This PR includes all fundamental endpoints for WizBox, which aims to provide suggestions for group hanging out.
- ```GET /boxes```: get all the boxes of one user
- ```POST /boxes```: create one new box
- ```POST /boxes/<int:box_id>/vote```: vote for all options in one box with 3 levels: 0 - sad, 1 - neutral, 2 - happy
- ```GET /boxes/<int:box_id>/options```: get all options of box_id
- ```POST /boxes/<int:box_id>/options```: create one new option/recommendation for box_id

minhmo1620 committed:
Add routes for wizbox

minhmo1620 committed:
Add endpoints to create new wizbox, vote options

minhmo1620 committed:
Add endpoints to add food options

minhmo1620 committed:
Create a table for wizboxes

minhmo1620 committed:
Create a table to store food option

minhmo1620 committed:
Create a table to store vote

minhmo1620 committed:
Add schemas

minhmo1620 committed:
Add coverage for pytest

minhmo1620 committed:
Fix the bug for option_id and update response value

minhmo1620 committed:
Add box_id validation, and update the response

minhmo1620 committed:
Update decode function by adding algorithms

minhmo1620 committed:
Add extra line by the end of the file

minhmo1620 committed:
Fix the argument name

minhmo1620 committed:
Add more helper functions for the unittest

minhmo1620 committed:
fix: Add vote as attribute of options table

minhmo1620 committed:
Add unittest for /boxes

minhmo1620 committed:
Add unittest for /options

minhmo1620 committed:
Add unittest for /vote

Ajaydip-Singh commented on README.md:
```diff
@@ -64,4 +64,8 @@ To run the unit tests use the following commands:
     $ pip install -r requirements.txt
-    $ pytest unit_test
\ No newline at end of file
+    $ pytest unit_test
```
You should add `venv_unit` to `.gitignore` to avoid accidentally committing it.

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+@token_required
+@validate_input(schema=BoxSchema)
+def create_new_box(user_id, data):
+    name = data['name']
+    description = data['description']
```
You should validate these fields more extensively. Currently, you are only doing type and length validation. 

With the current code, `"@#$#$#%^"` is a valid name for the box and obviously, such input does not make sense at all.

Something as simple as checking if the string is alphanumeric `.alnum()` or alphabetic `.alpha()` may be a good starting point.

minhmo1620 replied:
I thought about this! ```alnum()``` seems to be a good validation if we want to allow only alphabet and number. However, I think name and description of the box depend on the owner. It will be False if I have input "minh." due to the dot, this is bad as well. Hence, I decided to keep the current version.

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+    db.session.add(new_box)
+    db.session.commit()
+
+    return jsonify(BoxSchema().dump(new_box)), 200
```
A status code `201` is more appropriate here.

See [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201)

> The HTTP 201 Created success status response code indicates that the request has succeeded and has led to the creation of a resource. The new resource is effectively created before this response is sent back and the new resource is returned in the body of the message, its location being either the URL of the request, or the content of the Location header.

```suggestion
    return jsonify(BoxSchema().dump(new_box)), 201
```

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+
+@boxes_blueprint.route("/boxes", methods=["POST"])
+@token_required
+@validate_input(schema=BoxSchema)
```
I noticed that if you don't send anything in the request body, there is no useful error message to indicate which are the required fields for the endpoint. In such a situation, I was expecting a `400` status code but received a `500`. 

If you send an empty JSON object or omit the name or description fields in the JSON object, the API handles this properly.

minhmo1620 replied:
✅

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+    for option in all_options:
+        option_id = str(option.id)
+        if option_id not in data:
+            return jsonify({"message": "Please vote for all options"}), 400
```
Since there could be multiple options available in a box, the client may not know which options are available. Therefore, a more useful error message could be used. Specifically, the options that are required could be sent back in the response e.g. `Please vote for all options: 1, 2, 3, 4, 5 are all the required options.`

I also noticed that you can vote for an option even though it does not exist. Currently, it is causing no harm but technically it is an invalid request as you should not be able to send votes for an option that does not exist.

minhmo1620 replied:
I will add validation for the data to make sure that they cannot vote. For the error message, I think revealing the options_id is not a good choice for security purposes. I will keep the same error message.

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+    new_vote = VoteModel(user_id, json.dumps(data))
+    db.session.add(new_vote)
+    db.session.commit()
+    return jsonify({"message": "Vote successfully"}), 200
```
This seems like a typo. Also, status code `201` sees more appropriate here as well.
```suggestion
    return jsonify({"message": "Voted successfully"}), 201
```

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+    data = data['votes']
+
+    # Query all options of one box
+    all_options = db.session.query(OptionModel).filter(OptionModel.box_id == box_id).all()
```
Here you should handle the case when there are no options in a box. In such a scenario, the user should not be allowed to make the request at all and should receive an error. However, with the current implementation, the client can send a request to make a vote to an option that does not exist and receive a `200` status code which is wrong (to the client the response was successful).

minhmo1620 replied:
I will add validation to make sure they cannot vote not available options, then even when there is no option, they will return 400. Thanks for this case!

Ajaydip-Singh commented on app/controllers/boxes.py:
```diff
@@ -0,0 +1,63 @@
+@token_required
+@validate_input(schema=VoteSchema)
+def vote_options(user_id, box_id, data):
+    data = data['votes']
```
There is no validation if a box with the sent `box_id` exists. Currently, I can enter an invalid box id and receive a 200 status code that the vote was successful.

Ajaydip-Singh commented on app/models/boxes.py:
```diff
@@ -0,0 +1,16 @@
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(100), nullable=False)
+    description = db.Column(db.String(100), nullable=False)
+    user_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
```
Maybe you could call this `creater_id` so then it is clear what the role of the user is with respect to the box. But `user_id` is not bad either.

Ajaydip-Singh commented on app/models/boxes.py:
```diff
@@ -0,0 +1,16 @@
+
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(100), nullable=False)
+    description = db.Column(db.String(100), nullable=False)
```
There is a discrepancy between the max size of `description` here in the boxes model and in the boxes schema. 

In the boxes model, the max size of the string is 100 characters.
In the boxes schemas, the max size is 200 characters.

minhmo1620 replied:
Good catch!

Ajaydip-Singh commented on app/models/options.py:
```diff
@@ -0,0 +1,21 @@
+        self.description = description
+        self.box_id = box_id
+        self.user_id = user_id
+        self.vote = json.dumps([0, 0, 0])
```
I agree with the decision to serialize the list with JSON. More secure and convenient to work with than PickleType, although arguably a little slower.

minhmo1620 replied:
I will keep the original version to keep it simple.

Ajaydip-Singh commented on app/models/votes.py:
```diff
@@ -0,0 +1,13 @@
+class VoteModel(db.Model):
+    __tablename__ = "votes"
+
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
+    data = db.Column(db.String, nullable=False)
```
Maybe you should additionally store the `box_id` and `option_id` as well so there is a reference on where a particular vote was made by the user (this is just a suggestion as I do not really know how you are intending to use the votes).

Or you may choose to only store the `option_id` but with this method, if you ever want to know the `box_id` for the option, you will need to make an extra query.

minhmo1620 replied:
That's true that it is better to know who vote what. I will add #TODO to fix this later. At this stage, it is fine to keep the vote anonymous.

Ajaydip-Singh commented on app/schemas/boxes.py:
```diff
@@ -0,0 +1,9 @@
+
+class BoxSchema(Base):
+    name = fields.Str(required=True, validate=validate.Length(max=100, min=1))
+    description = fields.Str(required=True, validate=validate.Length(max=200, min=1))
```
See comment made regarding the discrepancy in the max size allowed for the `description` field in the boxes model.

minhmo1620 replied:
Fixed!

Ajaydip-Singh commented on app/controllers/options.py:
```diff
@@ -0,0 +1,38 @@
+@token_required
+@validate_input(schema=OptionSchema)
+def create_new_option(box_id, user_id, data):
+    name = data['name']
+    description = data['description']
```
You should validate these fields more extensively. Currently, you are only doing type and length validation.

With the current code, "@#$#$#%^" is a valid name for the option and obviously, such input does not make sense at all.

Something as simple as checking if the string is alphanumeric `.alnum()` or alphabetic `.alpha()` may be a good starting point.

minhmo1620 replied:
Same as above. I will keep the original version.

Ajaydip-Singh commented on app/controllers/options.py:
```diff
@@ -0,0 +1,38 @@
+    db.session.add(new_option)
+    db.session.commit()
+
+    return jsonify(OptionSchema().dump(new_option)), 200
```
Again, status code `201` is more appropriate here as well.

```suggestion
    return jsonify(OptionSchema().dump(new_option)), 201
```

Ajaydip-Singh commented on app/controllers/options.py:
```diff
@@ -0,0 +1,38 @@
+    return jsonify(OptionSchema(many=True).dump(list_options)), 200
+
+
+@options_blueprint.route("/boxes/<int:box_id>/options", methods=["POST"])
+@token_required
+@validate_input(schema=OptionSchema)
+def create_new_option(box_id, user_id, data):
```
I noticed that if you don't send anything in the request body, there is no useful error message to indicate which are the required fields for the endpoint. In such a situation, I was expecting a 400 status code but received a 500.

If you send an empty JSON object or omit the name or description fields in the JSON object, the API handles this properly.

minhmo1620 replied:
✅

Ajaydip-Singh commented on unit_test/test_boxes.py:
```diff
@@ -0,0 +1,71 @@
+from unit_test.helpers import create_dummy_user, create_headers, create_dummy_box
+
+
+def test_create_new_box(client):
+    """
+    Test: Create a new wizbox POST "boxes"
+    """
+    # create header
+    token = create_dummy_user(username="mia", password="abc")
+    headers = create_headers(token)
+
+    # invalid input
+    data = {"name": 1, "description": "Vietnamese food"}
+    response = client.post("/boxes", json=data, headers=headers)
+    assert response.status_code == 400
+    assert {"name": ["Not a valid string."]} == json.loads(response.data)
+
+    # invalid input
+    data = {"name": "Minerva Feast", "description": 1.2}
+    response = client.post("/boxes", json=data, headers=headers)
+    assert response.status_code == 400
+    assert {"description": ["Not a valid string."]} == json.loads(response.data)
+
+    # valid input
+    data = {"name": "Minerva Feast", "description": "06-02-2020"}
+    response = client.post("/boxes", json=data, headers=headers)
+    assert response.status_code == 200
+    expected_response = {
+        "name": "Minerva Feast",
+        "description": "06-02-2020",
+        "id": 1,
+    }
+    assert expected_response == json.loads(response.data)
```
Great tests to check the validity of the request body. 

To make tests more robust you should add tests for:

1. Empty JSON object as the request body.
2. No request body (send no object in the request body).

minhmo1620 replied:
✅

Ajaydip-Singh commented on unit_test/test_options.py:
```diff
@@ -0,0 +1,85 @@
+    headers = create_headers(token)
+    box_id = create_dummy_box(1, "Minerva Feast", "06-02-2021")
+
+    # invalid input
+    data1 = {"name": 1, "description": "Vietnamese food"}
+    response = client.post(f"/boxes/{box_id}/options", json=data1, headers=headers)
+    assert response.status_code == 400
+    assert {"name": ["Not a valid string."]} == json.loads(response.data)
+
+    # invalid input
+    data1 = {"name": "Ngon ngon", "description": 1}
+    response = client.post(f"/boxes/{box_id}/options", json=data1, headers=headers)
+    assert response.status_code == 400
+    assert {"description": ["Not a valid string."]} == json.loads(response.data)
+
+    # invalid box_id
+    data1 = {"name": "Ngon ngon", "description": "Vietnamese food"}
+    response = client.post("/boxes/10/options", json=data1, headers=headers)
+    assert response.status_code == 404
+    assert {'message': 'Cannot find the wizbox'} == json.loads(response.data)
+
+    # valid input
+    data1 = {"name": "Ngon ngon", "description": "Vietnamese food"}
+    response = client.post(f"/boxes/{box_id}/options", json=data1, headers=headers)
+    assert response.status_code == 200
+    expected_response = {
+        "name": "Ngon ngon",
+        "description": "Vietnamese food",
+        "id": 1
+    }
```
Great tests to check the validity of the request body.

To make tests more robust you should add tests for:

1. Empty JSON object as the request body.
2. No request body (send no object in the request body).

minhmo1620 replied:
✅

Ajaydip-Singh commented on unit_test/test_votes.py:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    response = client.post(f'/boxes/{box_id}/vote', json=data, headers=headers)
```
You need to add a test for invalid `box_id`.

Ajaydip-Singh reviewed (changes requested):
Overall great work! I have provided comments on the specific files to make it easier on what needs to be changed or what could be improved.

General comments:

1. Please add more comments. The bare minimum is to have comments for the controllers so that one can easily know what is required to make a successful request to the endpoint. Optionally, you might be interested in tools that automatically create API documentation from comments in source code (it might be overkill but if you are commenting code properly might as well have documentation for free).
2. In some situations a status code `201` seemed more appropriate.
3. Validation checks in POST requests should extend more than merely checking for the correct type and size.

minhmo1620 committed:
Add db and venv_unit to gitignore

minhmo1620 committed:
Change status code from 200 to 201

minhmo1620 committed:
fix: Change the message and status code from 200 to 201

minhmo1620 committed:
Add helpers within controllers

minhmo1620 committed:
Validate box_id and 404 if not found

minhmo1620 committed:
Validate options in vote api

minhmo1620 committed:
Add Todo for future fix

minhmo1620 committed:
Change user_id in BoxModel to owner_id

minhmo1620 committed:
Change the size for description to match with schema

minhmo1620 committed:
Add validation for body in request

minhmo1620 commented:
I fixed based on the feedbacks above. If the comment has no interaction, that means I fixed it. Otherwise, I commented my answers to decide to change the code or not.

minhmo1620 committed:
Add comments to the functions

Ajaydip-Singh commented:
Sounds great @minhmo1620! Feel free to merge then 😃

Ajaydip-Singh commented on app/helpers.py:
```diff
@@ -49,6 +49,9 @@ def decorator(f):
         def wrapper(*args, **kwargs):
             # take the input
             data = request.get_json()
+            print(data)
```
Mhmmm, is this a lingering print statement? Once this is addressed feel free to merge 😃 🚀

minhmo1620 replied:
Nice catch! Thank you!

minhmo1620 committed:
Delete print statement
